<?xml version="1.0"?>

<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:macro name="panda_gazebo" params="robot_name load_gripper">
      <gazebo reference="${robot_name}_joint7">
        <implicitSpringDamper>1</implicitSpringDamper>
        <provideFeedback>true</provideFeedback>
      </gazebo>
      <gazebo reference="${robot_name}_joint1">
        <implicitSpringDamper>1</implicitSpringDamper>
        <provideFeedback>true</provideFeedback>
      </gazebo>
      <gazebo reference="${robot_name}_joint2">
        <implicitSpringDamper>1</implicitSpringDamper>
        <provideFeedback>true</provideFeedback>
      </gazebo>
      <gazebo reference="${robot_name}_joint3">
        <implicitSpringDamper>1</implicitSpringDamper>
        <provideFeedback>true</provideFeedback>
      </gazebo>
      <gazebo reference="${robot_name}_joint4">
        <implicitSpringDamper>1</implicitSpringDamper>
        <provideFeedback>true</provideFeedback>
      </gazebo>
      <gazebo reference="${robot_name}_joint5">
        <implicitSpringDamper>1</implicitSpringDamper>
        <provideFeedback>true</provideFeedback>
      </gazebo>
      <gazebo reference="${robot_name}_joint6">
        <implicitSpringDamper>1</implicitSpringDamper>
        <provideFeedback>true</provideFeedback>
      </gazebo>
      <xacro:if value="${load_gripper}">
          <gazebo reference="${robot_name}_finger_joint1">
            <implicitSpringDamper>1</implicitSpringDamper>
            <provideFeedback>true</provideFeedback>
          </gazebo>
          <gazebo reference="${robot_name}_finger_joint2">
            <implicitSpringDamper>1</implicitSpringDamper>
            <provideFeedback>true</provideFeedback>
          </gazebo>
      </xacro:if>

    <!-- Adding FT sensor at wrist. This is not present in real robot and is used only for computing end-point effort directly, unlike in the real robot where it is estimated from the joint efforts. -->
    <gazebo reference="${robot_name}_joint7">
      <disableFixedJointLumping>true</disableFixedJointLumping>
      <sensor name="${robot_name}_joint7_force_torque" type="force_torque">
        <always_on>true</always_on>
        <visualize>false</visualize>
        <force_torque>
          <frame>sensor</frame>
          <measure_direction>child_to_parent</measure_direction>
        </force_torque>
      </sensor>
    </gazebo>
    <gazebo>
      <plugin name="ft_sensor" filename="libgazebo_ros_ft_sensor.so">
        <updateRate>30.0</updateRate>
        <topicName>/gazebo/robot/wrist/ft</topicName>
        <jointName>panda_joint7</jointName>
      </plugin>
    </gazebo>

    <!-- Configure self collision properties per link -->
    <!-- The gripper fingers do not have this flag otherwise the gripper would not be allowed to close correctly -->
    <gazebo reference="${robot_name}_link1">
      <selfCollide>true</selfCollide>
    </gazebo>

    <gazebo reference="${robot_name}_link2">
      <selfCollide>true</selfCollide>
    </gazebo>

    <gazebo reference="${robot_name}_link3">
      <selfCollide>true</selfCollide>
    </gazebo>

    <gazebo reference="${robot_name}_link4">
      <selfCollide>true</selfCollide>
    </gazebo>

    <gazebo reference="${robot_name}_link5">
      <selfCollide>true</selfCollide>
    </gazebo>

    <gazebo reference="${robot_name}_link6">
      <selfCollide>true</selfCollide>
    </gazebo>

    <gazebo reference="${robot_name}_link7">
      <selfCollide>true</selfCollide>
    </gazebo>
  
    <!-- Collision sensors on all links. This is not present on the real robot, it is used only in simulation to detect collisions. -->
    <!-- Link 0 is omitted because it is always in contact with the ground  -->
    <gazebo reference="${robot_name}_link1">
      <sensor name='${robot_name}_link1_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_link1_collision</collision>
          </contact>
          <plugin name="link1_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_link1_collision</bumperTopicName>
            <frameName>${robot_name}_link1</frameName>
          </plugin>
      </sensor>
    </gazebo>

    <gazebo reference="${robot_name}_link2">
      <sensor name='${robot_name}_link2_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_link2_collision</collision>
          </contact>
          <plugin name="link2_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_link2_collision</bumperTopicName>
            <frameName>${robot_name}_link2</frameName>
          </plugin>
      </sensor>
    </gazebo>

    <gazebo reference="${robot_name}_link3">
      <sensor name='${robot_name}_link3_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_link3_collision</collision>
          </contact>
          <plugin name="link3_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_link3_collision</bumperTopicName>
            <frameName>${robot_name}_link3</frameName>
          </plugin>
      </sensor>
    </gazebo>

    <gazebo reference="${robot_name}_link4">
      <sensor name='${robot_name}_link4_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_link4_collision</collision>
          </contact>
          <plugin name="link4_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_link4_collision</bumperTopicName>
            <frameName>${robot_name}_link4</frameName>
          </plugin>
      </sensor>
    </gazebo>

    <gazebo reference="${robot_name}_link5">
      <sensor name='${robot_name}_link5_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_link5_collision</collision>
          </contact>
          <plugin name="link5_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_link5_collision</bumperTopicName>
            <frameName>${robot_name}_link5</frameName>
          </plugin>
      </sensor>
    </gazebo>

    <gazebo reference="${robot_name}_link6">
      <sensor name='${robot_name}_link6_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_link6_collision</collision>
          </contact>
          <plugin name="link6_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_link6_collision</bumperTopicName>
            <frameName>${robot_name}_link6</frameName>
          </plugin>
      </sensor>
    </gazebo> 

    <gazebo reference="${robot_name}_link7">
      <sensor name='${robot_name}_link7_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_link7_collision</collision>
          </contact>
          <plugin name="link7_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_link7_collision</bumperTopicName>
            <frameName>${robot_name}_link7</frameName>
          </plugin>
      </sensor>
    </gazebo>

    <gazebo reference="${robot_name}_leftfinger">
      <sensor name='${robot_name}_leftfinger_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_leftfinger_collision</collision>
          </contact>
          <plugin name="leftfinger_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_leftfinger_collision</bumperTopicName>
            <frameName>${robot_name}_leftfinger</frameName>
          </plugin>
      </sensor>
    </gazebo>

    <gazebo reference="${robot_name}_rightfinger">
      <sensor name='${robot_name}_rightfinger_contact' type='contact'>
        <update_rate> 50 </update_rate>
        <always_on>true</always_on>
          <contact>
              <collision>${robot_name}_rightfinger_collision</collision>
          </contact>
          <plugin name="rightfinger_bumper" filename="libgazebo_ros_bumper.so">
            <bumperTopicName>${robot_name}_rightfinger_collision</bumperTopicName>
            <frameName>${robot_name}_rightfinger</frameName>
          </plugin>
      </sensor>
    </gazebo>

    </xacro:macro>

    <!-- camera -->
    <gazebo reference="camera_link">
      <sensor type="camera" name="camera1">
        <update_rate>30.0</update_rate>
        <camera name="head">
          <horizontal_fov>1.3962634</horizontal_fov>
          <image>
            <width>800</width>
            <height>800</height>
            <format>R8G8B8</format>
          </image>
          <clip>
            <near>0.02</near>
            <far>300</far>
          </clip>
          <noise>
            <type>gaussian</type>
            <!-- Noise is sampled independently per pixel on each frame.
                That pixel's noise value is added to each of its color
                channels, which at that point lie in the range [0,1]. -->
            <mean>0.0</mean>
            <stddev>0.007</stddev>
          </noise>
        </camera>
        <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
          <alwaysOn>true</alwaysOn>
          <updateRate>0.0</updateRate>
          <cameraName>rrbot/camera1</cameraName>
          <imageTopicName>image_raw</imageTopicName>
          <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          <frameName>camera_link_optical</frameName>
          <!-- setting hackBaseline to anything but 0.0 will cause a misalignment
              between the gazebo sensor image and the frame it is supposed to
              be attached to -->
          <hackBaseline>0.0</hackBaseline>
          <distortionK1>0.0</distortionK1>
          <distortionK2>0.0</distortionK2>
          <distortionK3>0.0</distortionK3>
          <distortionT1>0.0</distortionT1>
          <distortionT2>0.0</distortionT2>
          <CxPrime>0</CxPrime>
          <Cx>0.0</Cx>
          <Cy>0.0</Cy>
          <focalLength>0.0</focalLength>
        </plugin>
      </sensor>
    </gazebo>

</robot>

